src/cloud/auth.rs

```diff
@@ -0,0 +1,70 @@
+use std::time::Duration;
+
+use async_std::task;
+
+use crate::cloud::client::{cloud_config_file, CloudClient, CloudConfig};
+use crate::cloud::options;
+use crate::options::CloudOptions;
+use crate::portable::local::write_json;
+use crate::print;
+
+const AUTHENTICATION_WAIT_TIME: i32 = 5 * 60;
+
+#[derive(Debug, serde::Deserialize)]
+struct UserSession {
+    id: String,
+    token: Option<String>,
+    auth_url: String,
+}
+
+pub async fn login(_c: &options::Login, options: &CloudOptions) -> anyhow::Result<()> {
+    do_login(&CloudClient::new(options)?).await
+}
+
+pub async fn do_login(client: &CloudClient) -> anyhow::Result<()> {
+    let UserSession {
+        id,
+        auth_url,
+        token: _,
+    } = client
+        .post("auth/sessions/", serde_json::json!({ "type": "CLI" }))
+        .await?;
+    let link = format!("{}{}", client.base_url, auth_url);
+    log::debug!("Opening URL in browser: {}", link);
+    if open::that(&link).is_ok() {
+        print::prompt("Please complete the authentication in the opened browser.");
+    } else {
+        print::prompt("Please open this link in your browser and complete the authentication:");
+        print::success_msg("Link", link);
+    }
+    for _ in 0..AUTHENTICATION_WAIT_TIME {
+        let UserSession {
+            id: _,
+            auth_url: _,
+            token,
+        } = client.get(format!("auth/sessions/{}", id)).await?;
```

Comment:
Since we're retrying, I think errors should be printed and suppressed (i.e. intermediate network error shouldn't be a problem).